docs/how-to-guides/upgrade-etcd.md

```diff
@@ -0,0 +1,112 @@
+  - [Step 3: Upgrade etcd](#step-3-upgrade-etcd)
+  - [Step 4: Verify upgrade](#step-4-verify-upgrade)
+  - [Step 5: Verify using `etcdctl`](#step-5-verify-using-etcdctl)
+
+## Introduction
+
+[Etcd](https://etcd.io/) is the most crucial component of a Kubernetes cluster. It stores the cluster state.
+
+This document will provide step by step guide on upgrading etcd in Lokomotive.
+
+## Steps
+
+Repeat the following steps on all the controller node one node at a time.
+
+### Step 1: Find out the IP and SSH
+
+Find the IP of the controller node by visiting the cloud provider dashboard and ssh into it.
+
+```bash
+ssh core@<IP Address>
+```
+
+### Step 2: Create necessary directories with correct permissions
+
+Latest etcd (`>= v3.4.10`) necessitates the data directory permissions to be `0700`, accordingly change the permissions. Verify the permissions are changed to `rwx------`.
+
+```bash
+sudo chmod 0700 /var/lib/etcd/
+sudo ls -ld /var/lib/etcd/
+```
+
+If the node reboots, we need the right settings in place so that `systemd-tmpfile` service does not alter the permissions of the data directory. To make the changes made above persistent run the following command:
+
+```bash
+echo "d    /var/lib/etcd 0700 etcd etcd - -" | sudo tee /etc/tmpfiles.d/etcd-wrapper.conf
+```
+
+### Step 3: Upgrade etcd
+
+Run the following commands:
+
+> **NOTE**: Before proceeding to other commands, set the `etcd_version` variable to the latest etcd version.
+
+```bash
+export etcd_version=<latest etcd version e.g. v3.4.10>
+
+sudo sed -i "s,ETCD_IMAGE_TAG=.*,ETCD_IMAGE_TAG=${etcd_version}," \
+        /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf
+sudo systemctl daemon-reload
+sudo systemctl restart etcd-member
+```
+
+### Step 4: Verify upgrade
+
+Verify that the etcd service is in `active (running)` state:
+
+```bash
+sudo systemctl status --no-pager etcd-member
+```
+
+Run the following command to see logs of the process since the last restart:
+
+```bash
+sudo journalctl _SYSTEMD_INVOCATION_ID=$(sudo systemctl \
+              show -p InvocationID --value etcd-member.service)
+```
+
+> **NOTE**: Do not proceed with the upgrade of the rest of the cluster if you encounter any errors.
+
+Once you see the following log line, you can discern that the etcd daemon has come up without errors:
+
+```log
+etcdserver: starting server... [version: 3.4.10, cluster version: to_be_decided]
+```
+
+Once you see the following log line, you can discern that the etcd has rejoined the cluster without issues:
+
+```log
+embed: serving client requests on 10.88.81.1:2379
+```
+
+### Step 5: Verify using `etcdctl`
+
+We can use `etcdctl` client to verify the state of etcd cluster.
+
+> **NOTE**: Before proceeding to other commands, set the `no_of_controller_nodes` variable to the number of controller nodes in the cluster.
+
+```bash
+export no_of_controller_nodes=<no of controller nodes>
+
+# Find the endpoint of this node's etcd:
```

Comment:
Of this node? Isn't this a list of the 3 endpoints? How is the code getting the one for this node?

Isn't currently serveral endpoint*s*?

Reply:
Earlier due to the misunderstanding I thought certs were only on controller-0, now that we know there are certs on each node we can only look at that particular node.

Reply:
But is it doing something specific for this node? The code seems quite generic. Like, let's say you have 3 nodes, one down. Would the command say something about it?

Reply:
No it does say anything about other nodes.

Reply:
Do have an example of the output of that command in a cluster with 3 controllers? I think it should include endpoints for the 3 of them.

Or am I missing something?

Reply:
> Or am I missing something?

I think I did a poor job of explaining what the problem was, let me try again.

> Of this node? Isn't this a list of the 3 endpoints? How is the code getting the one for this node?

```console
$ cat /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf
[Unit]
Requires=create-etcd-config.service
After=create-etcd-config.service
[Service]
EnvironmentFile=/etc/kubernetes/etcd.config
Environment="ETCD_IMAGE_URL=docker://quay.io/coreos/etcd"
Environment="RKT_RUN_ARGS=--insecure-options=image"
Environment="ETCD_IMAGE_TAG=v3.4.7"
Environment="ETCD_NAME=etcd0"
Environment="ETCD_ADVERTISE_CLIENT_URLS=https://foobar-cluster-etcd0.k8s.com:2379"
Environment="ETCD_INITIAL_ADVERTISE_PEER_URLS=https://foobar-cluster-etcd0.k8s.com:2380"
Environment="ETCD_INITIAL_CLUSTER=etcd0=https://foobar-cluster-etcd0.k8s.com:2380,etcd1=https://foobar-cluster-etcd1.k8s.com:2380,etcd2=https://foobar-cluster-etcd2.k8s.com:2380"
Environment="ETCD_STRICT_RECONFIG_CHECK=true"
Environment="ETCD_SSL_DIR=/etc/ssl/etcd"
Environment="ETCD_TRUSTED_CA_FILE=/etc/ssl/certs/etcd/server-ca.crt"
Environment="ETCD_CERT_FILE=/etc/ssl/certs/etcd/server.crt"
Environment="ETCD_KEY_FILE=/etc/ssl/certs/etcd/server.key"
Environment="ETCD_CLIENT_CERT_AUTH=true"
Environment="ETCD_PEER_TRUSTED_CA_FILE=/etc/ssl/certs/etcd/peer-ca.crt"
Environment="ETCD_PEER_CERT_FILE=/etc/ssl/certs/etcd/peer.crt"
Environment="ETCD_PEER_KEY_FILE=/etc/ssl/certs/etcd/peer.key"
Environment="ETCD_PEER_CLIENT_CERT_AUTH=true"
Environment=""
ExecStopPost=-/opt/etcd-rejoin
```

Each etcd conf file has a list of all servers but on the peer port as in `2380`. While when we as client talk to this etcd server on port `2379`.

Earlier due to my misunderstanding of the certificates not being on all nodes I suggested the user(in the doc) to run this particular last step only on the first controller node. But later I found that the certs are available on all the nodes.

Now the earlier script I had was written such that it ran correctly only on the controller-0. It use to extract the endpoint of controller-0 and then generate a list by replacing the string `etcd0`. Now the same script won't work on all nodes because on controller-1 the endpoint has etcd1 and so on. Now the script there for user to copy and paste was becoming very complicated so I removed all that and now ask user to only check the health of the node that user is on.

Another way to do is to get the list of peers from the service file do replace of 2380 by 2379 and then remove things like etcd0=, etc1=, etc. I tried this approach as well, but again the complexity of it was defeating the purpose of keeping the document simplistic.

> Isn't currently serveral endpoints?

There are several endpoints in the service file but they are listed as shown in the file above.

> But is it doing something specific for this node? The code seems quite generic. Like, let's say you have 3 nodes, one down. Would the command say something about it?

The command tells you about all the endpoints we give it. We just have to include comma separated list of all the endpoints. But as explained above I have changed it to include only the current node.

> Do have an example of the output of that command in a cluster with 3 controllers? I think it should include endpoints for the 3 of them.

```console
$ sudo ETCDCTL_API=3 etcdctl member list $flags
2b89a567f6c2e2e8, started, etcd0, https://foobar-cluster-etcd0.k8s.com:2380, https://foobar-cluster-etcd0.k8s.com:2379
99588ac1031498cf, started, etcd2, https://foobar-cluster-etcd2.k8s.com:2380, https://foobar-cluster-etcd2.k8s.com:2379
eec240697239d5b8, started, etcd1, https://foobar-cluster-etcd1.k8s.com:2380, https://foobar-cluster-etcd1.k8s.com:2379
```

```console
$ sudo ETCDCTL_API=3 etcdctl endpoint health $flags
https://foobar-cluster-etcd0.k8s.com:2379 is healthy: successfully committed proposal: took = 30.131317ms
```

Reply:
Ohhh, I see. Sorry, I think that was partly my fault because the actual code line was not selected on my comment and that changed :facepalm: .

I now understand. I don't have all the details in mind about etcdctl, but I trust you and seems reasonable to me.

Thanks again for your patience! :)